### PR TITLE
[PLC] [Builtins] Added parametric polymorphism

### PR DESCRIPTION
**Pull request by effectfully**
This implements parametric polymorphism for builtins the way it's described [here](https://github.com/input-output-hk/plutus/issues/581#issuecomment-469294199). Works like a charm, the only annoying thing is that we need to explicitly provide uniques at the type level, but that won't be a problem with the new dynamic builtin types machinery.

Also enables type checking for dynamic builtin types tests.

### CODE REVIEW

File: language-plutus-core/prelude/PlutusPrelude.hs [hunk 1]
```diff
@@ -226,6 +227,11 @@ foldMapM :: (Foldable f, Monad m, Monoid b) => (a -> m b) -> f a -> m b
 foldMapM f xs = foldr step return xs mempty where
     step x r z = f x >>= \y -> r $! z `mappend` y
 
+-- | This function generalizes 'eitherToMaybe', 'eitherToList',
+-- 'listToMaybe' and other such functions.
+reoption :: (Foldable f, Alternative g) => f a -> g a
```
Comment (michaelpj) [inline language-plutus-core/prelude/PlutusPrelude.hs]: `Control.Applicative.Alternative.afold`?

File: language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs [hunk 2]
```diff
@@ -60,6 +60,17 @@ instance KnownDynamicBuiltinType a => KnownDynamicBuiltinType (EvaluationResult
     readDynamicBuiltin eval term =
         mapExceptT (EvaluationSuccess . sequence) <$> readDynamicBuiltin eval term
 
+instance KnownDynamicBuiltinType Int where
+    toTypeEncoding _ = TyApp () (TyBuiltin () TyInteger) (TyInt () 8)
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs]: Sad to see my integer sizing assumptions re-appearing here... not sure what else you can do atm.

File: language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs [hunk 3]
```diff
@@ -312,6 +340,15 @@ readDynamicBuiltinM
     => Term TyName Name () -> Evaluate m (Convert dyn)
 readDynamicBuiltinM term = withEvaluator $ \eval -> readDynamicBuiltin eval term
 
+-- | The denotation of a term whose type is a bound variable.
+-- I.e. the denotation of such a term is the term itself.
+-- This is because we have parametricity in Haskell, so we can't inspect a value whose
+-- type is a bound variable, so we never need to convert such a term from Plutus Core to Haskell
+-- and back and instead can keep it intact.
+newtype QuotedTerm (text :: Symbol) (unique :: Nat) = QuotedTerm
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs]: I think `QuotedTerm` is a little confusing considering `Quote`. Perhaps `AbstractTerm`, or `OpaqueTerm`?

File: plutus-core-interpreter/test/DynamicBuiltins/Definition.hs [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+        assertBool "type checks" $ isRight lhs
+        lhs @?= rhs
+
+dynamicConstName :: DynamicBuiltinName
+dynamicConstName = DynamicBuiltinName "const"
+
+dynamicConstMeaning :: DynamicBuiltinNameMeaning
+dynamicConstMeaning = DynamicBuiltinNameMeaning sch Prelude.const where
+    sch =
+        TypeSchemeAllType @"a" @0 Proxy $ \a ->
```
Comment (michaelpj) [inline plutus-core-interpreter/test/DynamicBuiltins/Definition.hs]: I was horrified and then I remembered we mostly only need well-scopedness, so this is okay and we'll rename it later? Is that true? Do we rename the types of dynamic builtins when we introduce them during typechecking?

File: plutus-core-interpreter/test/DynamicBuiltins/Definition.hs [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+        lhs === Right (Right (EvaluationSuccess c))
+        lhs === rhs
+
+dynamicReverseName :: DynamicBuiltinName
+dynamicReverseName = DynamicBuiltinName "reverse"
+
+dynamicReverseMeaning :: DynamicBuiltinNameMeaning
+dynamicReverseMeaning = DynamicBuiltinNameMeaning sch (PlcList . Prelude.reverse . unPlcList) where
+    sch =
+        TypeSchemeAllType @"a" @0 Proxy $ \(_ :: TypedBuiltin size a) ->
```
Comment (michaelpj) [inline plutus-core-interpreter/test/DynamicBuiltins/Definition.hs]: Hang on, shouldn't the argument to the lambda correspond to the bound type variable, i.e. the type of the elements of the list, not a size?

File: language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs [hunk 6]
```diff
@@ -131,11 +138,32 @@ data TypedBuiltinValue size a = TypedBuiltinValue (TypedBuiltin size a) a
 data TypeScheme size a r where
     TypeSchemeBuiltin :: TypedBuiltin size a -> TypeScheme size a a
     TypeSchemeArrow   :: TypeScheme size a q -> TypeScheme size b r -> TypeScheme size (a -> b) r
+    TypeSchemeAllType
+        :: (KnownSymbol text, KnownNat uniq)
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs]: I get it, I think, but this is still a bit scary :fearful:

File: language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs [hunk 7]
```diff
@@ -312,6 +340,15 @@ readDynamicBuiltinM
     => Term TyName Name () -> Evaluate m (Convert dyn)
 readDynamicBuiltinM term = withEvaluator $ \eval -> readDynamicBuiltin eval term
 
+-- | The denotation of a term whose type is a bound variable.
+-- I.e. the denotation of such a term is the term itself.
+-- This is because we have parametricity in Haskell, so we can't inspect a value whose
+-- type is a bound variable, so we never need to convert such a term from Plutus Core to Haskell
+-- and back and instead can keep it intact.
```
Comment (michaelpj) [inline language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs]: I like this explanation :+1:

File: language-plutus-core/prelude/PlutusPrelude.hs (hunk 1 above)
Comment (effectfully) [reply language-plutus-core/prelude/PlutusPrelude.hs]: Wasn't aware of that. It takes more indirections, though:

```
afromList :: Alternative f => [a] -> f a
afromList = Foldable.asum . map pure

afold :: (Alternative f, Foldable t) => t a -> f a
afold = afromList . Foldable.toList
```

I like the direct implementation more.

Comment (michaelpj) [reply language-plutus-core/prelude/PlutusPrelude.hs]: I also like the direct implementation more, but I like using library functions more :)

File: language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs (hunk 2 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs]: Not making any particular assumptions here, because the instance is for `Int` and not `Integer`. 

Strictly speaking, `Int` is platform-dependent, so we may want to use `Data.Bits.finiteBitSize` instead, but 8 bytes seems like a good initial implementation. Do you think we should use `finiteBitSize`?

File: language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs (hunk 3 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs]: `OpaqueTerm` is great, thanks, I'll change the name.

File: plutus-core-interpreter/test/DynamicBuiltins/Definition.hs (hunk 5 above)
Comment (effectfully) [reply plutus-core-interpreter/test/DynamicBuiltins/Definition.hs]: It does correspond to the type of the elements of the list. The `size` variable is bound only because we can't really omit it without enabling some extension. I can alias `TypedBuiltin size` to `bound` the same way I alias `QuotedTerm text uniq` to `a`, so that you can write

```
TypeSchemeAllType @"a" @0 Proxy $ \(_ :: bound a) ->
```

but I'm not sure if I should do that, because in the `const` test above we use the fact that the bound term-level variable is a `TypedBuiltin`, so another alias may look misleading (even though it does not break anything).

File: language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs (hunk 6 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Constant/Typed.hs]: The new constructor, a bit surprisingly, does not introduce any complication at all. You basically never need to think how to instantiate the continuation, because it always gets instantiated by the same `TypedBuiltinDyn` (just do `Ctrl-f `TypeSchemeAllType` on the changes tab and see yourself). So what worked previously, still works, and what didn't work (`error "Not implemented."`) still does not work.

It won't always be like that, though, because right now we manually provide type uniques, which is silly, and once we stop doing that, there will be corner cases, for example you won't be able to always compute the resulting type of a `TypeScheme`, because as a standalone type it may contain a free variable (we need this only for tests and do not really care much), but I do expect the final design to be nice to deal with.

File: language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs (hunk 2 above)
Comment (michaelpj) [reply language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs]: No, I think we should either have properly sized integers on the Haskell side or get rid of sized integers in PLC :)

File: plutus-core-interpreter/test/DynamicBuiltins/Definition.hs (hunk 4 above)
Comment (effectfully) [reply plutus-core-interpreter/test/DynamicBuiltins/Definition.hs]: > Do we rename the types of dynamic builtins when we introduce them during typechecking?

Yes. When turning `DynamicBuiltinNameMeanings` into `DynamicBuiltinNameTypes`, we normalize the types along the way using the safe interface of the type normalizer which renames its argument first.

We also always explicitly apply functions to those bound type variables and never inline them. E.g.

```haskell
instance KnownDynamicBuiltinType a => KnownDynamicBuiltinType (PlcList a) where
    toTypeEncoding _ = TyApp () (_recursiveType listData) $ toTypeEncoding (Proxy @a)
```

This ensures that a free variable won't become bound by an unexpected binder. And the renamer handles name shadowing issues.

The current assumption we have is that only `QuotedTerm` is allowed to have a non-closed type and all other `KnownDynamicBuiltinType` must be closed and uniqueness of type variables in `QuotedTerm` is managed by the user. This looks like a pretty safe design to me (modulo explicit uniques managing, but I'm going to fix that at some point).

File: language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs (hunk 2 above)
Comment (effectfully) [reply language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs]: Yes, I think the same, but again, it's `Int`, not `Integer`, it does have an upper bound (even though it's platform-dependent).

File: language-plutus-core/prelude/PlutusPrelude.hs (hunk 1 above)
Comment (effectfully) [reply language-plutus-core/prelude/PlutusPrelude.hs]: Okay, my trump card: I like the `reoption` name more, because it indicates what is happening: a transformation from one type treated as optional to another type treated the same way.

Note also that `afold` behaves differently: `asum` calls `(<|>)`, so in the end we have `pure x <|> r`, but this is only equal to `x` for alternatives that obey the left catch law and many collections do not obey it, so for example using `afold` over a tree is going to return all the elements of a tree in a, say, list. While `reoption` would return the head of the tree as a singleton list.

I think both operation are sensible and useful, but right now I'm using only the option types mapping behavior, so I prefer to keep `reoption`.

**Comment by effectfully** [comment]
Merging.
